ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java

```diff
@@ -453,6 +462,40 @@ private void cleanupOldState() {
     localCommands.ifPresent(lc -> lc.processLocalCommandFiles(serviceContext));
   }
 
+  public void cleanupOldStateDirectories(final KsqlConfig configWithApplicationServer) {
+    final String stateDir =
+        configWithApplicationServer
+            .getKsqlStreamConfigProps()
+            .get(StreamsConfig.STATE_DIR_CONFIG)
+            .toString();
+
+    final Set<String> stateStoreNames =
+        ksqlEngine.getPersistentQueries()
+            .stream()
+            .map(PersistentQueryMetadata::getQueryId)
+            .map(QueryId::toString)
+            .collect(Collectors.toSet());
+    try {
+      Files.list(Paths.get(stateDir))
+          .map(Path::toFile)
+          .forEach(
+              f -> {
+                if (stateStoreNames.stream().noneMatch((name) -> f.getName().endsWith(name))) {
+                  try {
+                    Files.walk(f.toPath())
```

Comment:
nit: I personally like the `Files.walkFileTree` since it allows us to print a more detailed error message, e.g. which file delete encounters the error, with the visitor pattern.

An example of its usage: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/utils/Utils.java#L812

Reply:
Plus, since we would not expect to be finding any existing state store dirs that do not match existing persistent stores, we'd also log a warning if we indeed successfully delete something here.. e.g. "WARN: Deleted local state store for non-existing query {}, this is not expected and probably due to a race condition when query {} was dropped before".